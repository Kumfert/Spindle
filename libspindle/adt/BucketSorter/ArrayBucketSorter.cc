//
// ArrayBucketSorter.cc
//
//  $Id: ArrayBucketSorter.cc,v 1.2 2000/02/18 01:31:43 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/ArrayBucketSorter.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

ArrayBucketSorter::~ArrayBucketSorter() {
  decrementInstanceCount( ArrayBucketSorter::MetaData );
  delete[] bucket; 
  delete[] next;   
  delete[] prev;
  delete[] key;
}

void 
ArrayBucketSorter::validate() {
  if ( currentState == VALID ) { return;}
  // else do some checking...
}

bool 
ArrayBucketSorter::insertBack(const int Key, const int Item, const int bucketID) {
  int head;
  if ((head = bucket[bucketID]) == FLAG) {    // if bucket is empty
    bucket[bucketID] = Item;                   //    set bucket to object
    prev[Item] = next[Item] = Item;       //    set double linked list
  } else {                                // else (bucket non-empty)
    next[Item] = head;                    //    insert at end of list
    prev[Item] = prev[head];
    next[prev[head]] = Item;
    prev[head] = Item;
  }                                       // endif
  key[Item] = Key;
  return true;                            // insert succeeded
}


bool 
ArrayBucketSorter::insertFront(const int Key, const int Item, const int bucketID) {
  int head;
  if ((head = bucket[bucketID]) == FLAG) {    // if bucket is empty
    bucket[bucketID] = Item;                   //    set bucket to object
    prev[Item] = next[Item] = Item;       //    set double linked list
  } else {                                // else (bucket non-empty)
    next[Item] = head;                    //    insert at end of list
    prev[Item] = prev[head];
    next[prev[head]] = Item;
    prev[head] = Item;
    bucket[bucketID] = Item;
  }                                       // endif
  key[Item] = Key;
  return true;                            // insert succeeded
}

bool
ArrayBucketSorter::removeItem(const int Item) {
  int myKey=key[Item];
  if ( myKey == FLAG ) {               // if object is unowned
    return false;                       //    removal failed
  } else {                              // else 
    int myBucket = myKey % nBucketsMax;
    if (next[Item] == Item) {           //    if sole object in bucket
      bucket[myBucket] = FLAG;         //       empty bucket
    } else {                            //    else
      if (bucket[myBucket] == Item) {   //       if top of list
	bucket[myBucket] = next[Item];  //          bucket starts next
      }                                 //       endif
      next[prev[Item]] = next[Item];    //       remove from rest of list
      prev[next[Item]] = prev[Item];
    }                                   //    endif
    key[Item] = FLAG;                //    reset Item's pointers
    next[Item] = FLAG;
    prev[Item] = FLAG;
    return true;                         //    delete succeeded
  }                                      // endif
}

int 
ArrayBucketSorter::nextItem(const int Item) const {
  int Key = key[Item];
  int bucketID = Key % nBucketsMax;
  if (Key == FLAG) {                    // if item not in buckets
    return FLAG;                        //    no next
  } else { 
    int next_Item=next[Item];   
    if (next_Item == bucket[bucketID]) { // if next item is head
      return FLAG;                      //    no next
    } else {
      return next_Item;                  // return next
    }
  }
}

int 
ArrayBucketSorter::prevItem(const int Item) const {
  int Key = key[Item];
  int bucketID = Key % nBucketsMax;
  if (Key == FLAG) {                    // if item not in buckets
    return FLAG;                        //    no next
  } else { 
    if ( Item == bucket[bucketID]) {    // if item is top of bucket
      return FLAG;                      //    no previous
    } else {
      return prev[Item];
    }
  }
}

bool
ArrayBucketSorter::reset() {
  currentState = VALID;
  nInserted=0;
  setInsertFront();
  int *ip;
  ip = bucket; {for (int i=0; i<nBucketsMax ;  ++i) *ip++ = FLAG;}
  ip = next;   {for (int i=0; i<nItemsMax;  ++i) *ip++ = FLAG;}
  ip = prev;   {for (int i=0; i<nItemsMax;  ++i) *ip++ = FLAG;}
  ip = key;    {for (int i=0; i<nItemsMax;  ++i) *ip++ = FLAG;}
  return true;
}


void 
ArrayBucketSorter::printDebug(FILE * fp ) const {
  fprintf(fp,"\nnBucketsMax = %d    nItemsMax = %d\n", nBucketsMax, nItemsMax );
  fprintf(fp,"    nInserted = %d\n", nInserted );

  fprintf(fp,"bucket: ");
  {for (int i=0; i<nBucketsMax; i++ ) { fprintf(fp,"%3d  ", bucket[i] ); }}
  
  fprintf(fp,"\nnext:   ");
  {for (int i=0; i<nItemsMax; i++ ) { fprintf(fp,"%3d  ", next[i] ); }}

  fprintf(fp,"\nprev:   ");
  {for (int i=0; i<nItemsMax; i++ ) { fprintf(fp,"%3d  ", prev[i] ); }}
  
  fprintf(fp,"\nkey:    ");
  {for (int i=0; i<nItemsMax; i++ ) { fprintf(fp,"%3d  ", key[i] ); }}
  fprintf(fp,"\n");
}

void 
ArrayBucketSorter::dump( FILE * fp ) const {
  fprintf(fp,"\nnInserted = %d\n", nInserted );
  for( int i=0; i < nBucketsMax; i++) {
    fprintf(fp,"%5d : ",i);
    if (bucket[i] != FLAG) {
      int cur = bucket[i];
      do{
	fprintf(fp," %3d ", cur );
	cur = next[cur];		    
      } while ( cur != bucket[i] );
    }
    fprintf(fp,"\n");
  }
}

void
ArrayBucketSorter::storeObject( SpindleArchive& ar ) const {
  ar << nBucketsMax; 
  ar << nItemsMax;
  ar << nInserted;
  ar << isInsertFront;
  // a better way to do this (rather than storing all arrays)
  // would be to simply store (item,key) pairs.
  pair< const int*, size_t > bucket_array( bucket, (size_t)nBucketsMax );
  pair< const int*, size_t > next_array( next, (size_t)nItemsMax );
  pair< const int*, size_t > prev_array( prev, (size_t)nItemsMax );
  pair< const int*, size_t > key_array( key, (size_t)nItemsMax );  
  ar << bucket_array;
  ar << next_array;
  ar << prev_array;
  ar << key_array;
}

void
ArrayBucketSorter::loadObject( SpindleArchive& ar ) {
  ar >> nBucketsMax;
  ar >> nItemsMax;
  ar >> nInserted;
  ar >> isInsertFront;

  bucket = new int[nBucketsMax];
  next   = new int[nItemsMax];
  prev   = new int[nItemsMax];
  key    = new int[nItemsMax];

  // A better way to do this would be to reset the sorter
  // at this point and read in (item,key) pairs, inserting
  // to end
  pair< int*, size_t > bucket_array( bucket, (size_t)nBucketsMax );
  pair< int*, size_t > next_array( next, (size_t)nItemsMax );
  pair< int*, size_t > prev_array( prev, (size_t)nItemsMax );
  pair< int*, size_t > key_array( key, (size_t)nItemsMax );
  
  ar >> bucket_array;
  ar >> next_array;
  ar >> prev_array;
  ar >> key_array;
}

SPINDLE_IMPLEMENT_PERSISTANT( ArrayBucketSorter, SpindlePersistant );

/*
#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

main() {
    ArrayBucketSorter bs(10,10);
    bs.printDebug();
    bs.insert(2,0);
    bs.insert(2,2);
    bs.insert(2,4);
    bs.insert(2,6);
    bs.insert(2,8);
    bs.printDebug();
    bs.print();
}
*/
