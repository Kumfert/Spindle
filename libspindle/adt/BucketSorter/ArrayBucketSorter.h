//
// ArrayBucketSorter.h
//
//  $Id: ArrayBucketSorter.h,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
//=====================================================================
//
// This bucket sorter can sort `nItemsMax' items into `nBucketsMax' buckets
// using a key which is hashed into the buckets (usually modulo nBucketsMax).
// Each Item must be unique, but the key (obviously) may not.
//
// Each bucket contains a circular doubly linked list.  Inserts are made 
// at the front or back of each list.  The bracket operator `[Key]' returns 
// the Item at the top of that bucket, or -1 if the bucket is empty.
//
// Items can be removed from anywhere in the ArrayBucketSorter in O(1).  This was
// a design constraint since this object is used in a Multiple Minimum Degree
// code.  This is also a reason why the Items must be unique.
//
// The user can query what the original key was, since the key is not
// necessarily equal to the bucket number.

#ifndef SPINDLE_ARRAY_BUCKET_SORTER_H_
#define SPINDLE_ARRAY_BUCKET_SORTER_H_

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE 

/**
 * @memo sorts 'nItemsMax' items with 'nBucketsMax' unique keys.
 * @type class
 *
 * @author Gary Kumfert
 */
class ArrayBucketSorter : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( ArrayBucketSorter )

protected:
  enum{ FLAG = -1 };    // C++ enum hack for a constant flag  
  /// number of unique keys 
  int nBucketsMax;    

  /// number of unique item ID numbers
  int nItemsMax;   
  int isInsertFront;

  /// Simple counter of number Items in ArrayBucketSorter
  int nInserted;         
  
  /// array of heads for each list length==(nBucketsMax)
  int *bucket;           

  /// array of forward pointers length==(nItemsMax)
  int *next;             
  
  /// array of back pointers length==(nItemsMax)
  int *prev;            

  /// array mapping each vertex in structure to its bucket
  int *key;              
  //    length==(nItemsMax)
public:
  
  /**
   * default constructor
   */
  ArrayBucketSorter();

  /** 
   * constructor
   */
  ArrayBucketSorter(const int nBuckets, const int nItems);

  /**
   * destructor
   */
  virtual ~ArrayBucketSorter();

  /**
   *
   */
  virtual void validate();

  /**
   *
   */
  virtual bool isValid() const;

  /**
   *  pretty prints the bucket structure
   */
  virtual void dump( FILE * stream ) const;

  /**
   *  pretty prints the bucket structure
   */
  void dump() const; 

  /**
   * @return true if there are no more Items to be removed
   */
  bool isEmpty() const;    

  /**
   * get the number inserted
   */
  int size() const;        

  /**
   * @return the minimum key of a non-empty bucket, -1 if none.
   */
  int queryMinBucket() const;  

  /**
   * @return minimum key of next non-empty bucket, -1 if none.
   */
  int queryNextBucket(const int Bucket) const;
  
  /**
   * @return the first item in the requested bucket, 
   *         -1 if out of range or bucket is empty
   */ 
  int queryFirstItem(const int Bucket) const;

  /**
   * @return the next item in the same bucket
   *         -1 if out of range or Item is last in bucket
   */
  int queryNextItem(const int Item) const; 

  /**
   * @return prev item in same bucket 
   *         -1 if out of range or Item is first in bucket
   */
  int queryPrevItem(const int Item) const;
  
  /**
   * @return the key of an item
   */
  int queryKey( const int Item );   

  /**
   * @return the bucket an item is in
   */
  int queryBucket( const int Item ); 

  /**
   * @return the number of items inserted into the bucketsorter
   */
  int queryNInserted() const;

  /**
   * @return true if insert was successful, 
   *         false if item was already in some bucket
   */
  bool insert(const int Key, const int Item);

  /** 
   * make all inserts occur to the front of the bucket
   */
  void setInsertFront();
  
  /**
   * make all inserts occur in the back of a bucket
   */
  void setInsertBack();
    
  /**
   * @return true if item is found and removed, 
   *         false if item is not found
   */
  bool remove(const int Item);
  
  /**
   * resets the ArrayBucketSorter to its initial (empty) state
   */
  bool reset();
  
  /**
   * print out complete internal state
   */
  void printDebug(FILE * fp) const;

  /**
   * print out complete internal state
   */
  void printDebug() const;
  
protected:
  /**
   * insert 'Item' with 'Key' into the front of bucket 'bucketID'
   * all arguments are assumed in range
   * @return true if insert succeeds, false otherwise
   */
  bool insertFront(const int Key, const int Item, const int bucketID);

  /**
   * insert 'Item' with 'Key' into the back of bucket 'bucketID'
   * all arguments are assumed in range
   * @return true if insert succeeds, false otherwise
   */
  bool insertBack(const int Key, const int Item, const int bucketID);

  /**
   * remove an 'Item' (assumed to be in range)
   * @return true if insert succeeds, false otherwise
   */
  bool removeItem(const int Item);

  /**
   * @return next item if it exists, -1 otherwise
   */
  int nextItem(const int Item) const;

  /**
   * @return next item if it exists, -1 otherwise
   */
  int prevItem(const int Item) const;
};

inline ArrayBucketSorter::ArrayBucketSorter() 
: nBucketsMax(0), nItemsMax(0){
  currentState = EMPTY;
}
  

inline ArrayBucketSorter::ArrayBucketSorter(const int nBucketsMax_, const int nItems) 
  : nBucketsMax(nBucketsMax_), nItemsMax(nItems) {
    incrementInstanceCount( ArrayBucketSorter::MetaData );
    bucket = new int[nBucketsMax];  
    next   = new int[nItemsMax]; 
    prev   = new int[nItemsMax]; 
    key    = new int[nItemsMax]; 
    reset();
}

inline bool
ArrayBucketSorter::isValid() const {
  return (currentState == VALID);
}

inline void 
ArrayBucketSorter::setInsertFront() {
  isInsertFront = true; 
}

inline void 
ArrayBucketSorter::setInsertBack() { 
  isInsertFront = false; 
}

inline bool 
ArrayBucketSorter::insert(const int Key, const int Item) { 
  if ((Item < 0) || (Item >= nItemsMax)) { return false; } // out of range
  if ( key[Item] != FLAG ) { return false; } // already inserted
  int bucketID = ( Key < nBucketsMax ) ? Key : Key % nBucketsMax ;
  bool insertSucceeded = (isInsertFront) ? insertFront(Key,Item,bucketID) : insertBack(Key,Item,bucketID);
  if (insertSucceeded) { ++nInserted; }
  return insertSucceeded;
}
  
// returns true if item is found and removed, false if item is not found
inline bool 
ArrayBucketSorter::remove(const int Item){
  bool removeSucceeded = ((Item < 0) || (Item >= nItemsMax)) ? false : removeItem(Item);
  if (removeSucceeded) { --nInserted; }
  return removeSucceeded;
}
  
inline bool 
ArrayBucketSorter::isEmpty() const { 
  return (nInserted == 0); 
}

inline int 
ArrayBucketSorter::queryKey( const int Item ) {
  if ( (Item<0) || (Item>nItemsMax) ) { 
    return FLAG; 
  } else {
    return key[Item];
  }
}


inline int 
ArrayBucketSorter::queryBucket( const int Item ) {
  if ( (Item<0) || (Item>nItemsMax) ) { 
    return FLAG; 
  } else {
    return key[Item] % nBucketsMax ;
  }
}

inline int 
ArrayBucketSorter::queryNInserted() const { 
  return nInserted; 
}

inline int 
ArrayBucketSorter::queryMinBucket() const { 
  if (isEmpty()) return FLAG;
  int i=0;
  while ((i<nBucketsMax) && (bucket[i] == FLAG)) ++i;
  return ( i == nBucketsMax ) ? FLAG : i;
}
  
// returns the minimum non-empty bucket, -1 if empty
#define __FUNC__ "Array ArrayBucketSorter::queryNextBucket(const int Bucket) const"
inline int 
ArrayBucketSorter::queryNextBucket(const int Bucket) const {       
  FENTER;
  if (isEmpty()) FRETURN( FLAG );
  ASSERT( Bucket>=0 , "Requested Bucket %d is invalid, must be positive", Bucket );
  ASSERT( Bucket<nBucketsMax, "Requested Bucket %d is invalid, number of Buckets=%d", Bucket,nBucketsMax);
  int i=Bucket;
  while ((i<nBucketsMax) && (bucket[i] == FLAG)) ++i;
  FRETURN( ( i == nBucketsMax ) ? FLAG : i );
}
#undef __FUNC__
  
inline int 
ArrayBucketSorter::queryFirstItem(const int Key) const { 
  return ( (Key >= 0) && ( Key < nBucketsMax) ) ? bucket[Key] : FLAG ; 
}

inline int 
ArrayBucketSorter::queryNextItem(const int Item) const {
  return ((Item<0) || (Item>=nItemsMax)) ? FLAG : nextItem(Item) ;
}  
  
inline int 
ArrayBucketSorter::queryPrevItem(const int Item) const {
  return ((Item<0) || (Item>=nItemsMax)) ? FLAG : prevItem(Item) ;
}  
  
inline void 
ArrayBucketSorter::dump() const{
  dump( stdout ); 
}
  
inline int  
ArrayBucketSorter::size() const {
  return nInserted; 
}


inline void  
ArrayBucketSorter::printDebug() const { 
  printDebug( stdout ); 
}

SPINDLE_END_NAMESPACE

#endif
