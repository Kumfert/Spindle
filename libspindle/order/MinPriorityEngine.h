//
// MinPriorityEngine.h
//
// $Id: MinPriorityEngine.h,v 1.2 2000/02/18 01:31:53 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef MIN_PRIORITY_ENGINE_H_
#define MIN_PRIORITY_ENGINE_H_

#ifdef DEBUG_TRACE
#undef DEBUG_TRACE
#endif

#define DEBUG_TRACE true

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_ORDERING_ALGORITHM_H_
#include "spindle/OrderingAlgorithm.h"
#endif

#ifndef SPINDLE_AUTO_PTR_H_
#include "spindle/SharedPtr.h"
#endif

#ifndef SPINDLE_PERMUTATION_H_
#include "spindle/PermutationMap.h"
#endif 

#ifndef SPINDLE_GRAPH_H_
#include "spindle/Graph.h"
#endif


#ifndef SPINDLE_QUOTIENT_GRAPH_H_
#include "spindle/QuotientGraph.h"
#endif

#ifndef SPINDLE_ARRAY_BUCKET_SORTER_H_
#include "spindle/ArrayBucketSorter.h"
#endif

SPINDLE_BEGIN_NAMESPACE

// Abstract Base Class for various priority strategiess.
// This will be defined below along with several 
// derived classes for external degree and approximate degree
class MinPriorityStrategies;

/**
 * @memo produced fill reducing orderings
 * @type class
 *
 * @author Gary Kumfert
 * @version #$Id: MinPriorityEngine.h,v 1.2 2000/02/18 01:31:53 kumfert Exp $#
 */
class MinPriorityEngine : public OrderingAlgorithm {
public:
  /**
   * used for moving sets of vertices around
   */
  typedef QuotientGraph::VertexList VertexList;

  /**
   * used for the priority queue
   */
  typedef ArrayBucketSorter BucketSorter;

private:
  int n; // original number of supernodes in graph
  bool randomizeGraph; // true iff we want to randomize the graph
  BucketSorter *degStruct; // The priority queue of (vertex,priority) pairs.
  QuotientGraph *qgraph;   // The quotient graph
  MinPriorityStrategies *priorityStrategy; // the priority strategy
  
  VertexList reachableVtxs;  // the list of reachable vertices
  VertexList mergedVtxs;     // the list of merged vertices
  // VertexList delayedVtxs; // the list of delayed vertices.
  VertexList workVtxList;    // a work list of vertices
  vector< pair< int, int > > purgedVtxs;  // actually degree, vtx pairs

  int delayWeight;  // 
  int delta;        // 
  int maxSteps;     // maximum number of eliminated supernodes for each update
  int *workVec;     // temporary work space
  float heavyVtxTolerance; // tolerance for removing highly connected vtxs 
  int * explicitOrder; // an explicit input ordering new2old (optional)


  //  SharedArray<int> PartitionMap; 
  stopwatch timer;
  stopwatch stageTimer;
  stopwatch stepTimer;
  stopwatch degreeComputationTimer;
  stopwatch delayedVtxsTimer;

  void processDelayedVtxs(); 
  bool initialize();
  bool run();
  bool finalize();

public:  
  /** default constructor */
  MinPriorityEngine();
  /** convenience constructor */
  MinPriorityEngine( const Graph* graph );
  /** destructor */
  virtual ~MinPriorityEngine();

  /** compute the ordering */
  virtual bool execute();
  /** reset class to a state of default constructor */
  virtual bool reset();

  /** set the graph to order */
  bool setGraph( const Graph * graph );
  /** set the priority strategy */
  bool setPriorityStrategy( MinPriorityStrategies* PriorityStrategy );
  /** force the input graph to be randomized */
  void setRandomizeGraph( bool randomizeIt ) { randomizeGraph = randomizeIt; }
  /** set an explicit input ordering for the graph */
  bool setExplicitOrder( const PermutationMap& perm );
  /** return a const pointer to current priority strategy */
  const MinPriorityStrategies* getPriorityStrategy();

  /** check if the entire quotient graph is eliminated */
  bool isAllPermuted() const { 
    for( int i=0; i<n; i++) {
      if ( !qgraph->isEliminated(i) ) {
	return false;
      }
    }
    return true;
  }
  
  /** delta is a tolerance for multiple elimination schemes */
  bool setDelta( const int i );
  /** query delta */
  int getDelta() const { return delta; }
  
  /** 
   * allows restriction of the maximum number of supernodes to
   * eliminate before updating the quotient graph.
   */
  bool setMaxSteps( const int i ) { 
    if ((i<0)||(i>n)) return false; 
    maxSteps = i; 
    return true; 
  }
  /** query the maximum number of supernodes eliminated per update */
  int getMaxSteps() const { return maxSteps; }
  
  //SharedPtr<PermutationMap> getPermutationMap() const;

  /**@name timer access */
  //@{
  /** const access to basic timer */
  const stopwatch& getTimer() const { return timer; }
  /** const access to timer that runs once per mass elimination */
  const stopwatch& lendStageTimer() const { return stageTimer; }
  /** const access to timer that runs one per eliminated supernode */
  const stopwatch& lendStepTimer() const { return stepTimer; }
  /** const access to timer that runs once per degree computation */
  const stopwatch& lendDegreeComputationTimer() const { return degreeComputationTimer; }
  /** const access to timer that tracks time spent on delayed Vtxs */
  const stopwatch& lendDelayedVtxsTimer() const { return delayedVtxsTimer; }
  //@}

  /** grant const access to the quotient graph itself */
  const QuotientGraph& lendQuotientGraph() const { return *qgraph; }
  /** (deprecated) reproduces an off by one bug in GENMMD */
  void enableIncrementByOne() { qgraph->enableIncrementByOne(); }
  /** (default) do not increment snodes adjacent to enodes by one */
  void disableIncrementByOne() { qgraph->disableIncrementByOne(); }

  // These routines are useful for indefinite problems
  //  bool enableDelayedVertices( const VertexList& delayedVertices );
  //  bool disableDelayedVertices();
  //  void setDelayWeight( const int i ) { delayWeight = i; }
  
  /** set tolerance for highly connected vertices to be pre-removed */
  bool setHeavyVertexTolerance( float tolerance ); // default 0.8

  // some pass-through settings for qgraph
  //  bool setMixAdjCompression( const int option );
  //  bool setNAdjClqCompressionBound( const int bound );
  //  int queryNPurgedVtxs() const ;
  // Partition Restrictions
  //  bool disablePartitionRestrictions() { return false; } // default
  //  bool enablePartitionRestrictions( const int partitionMap[] ) { return false; }

  SPINDLE_DECLARE_DYNAMIC( MinPriorityEngine )
};

/*
inline bool 
MinPriorityEngine::setMixAdjCompression( const int option ) {
  if ( qgraph != 0 ) { 
    return qgraph->setMixAdjCompression( option );
  } else { 
    return false;
  }
}

inline bool 
MinPriorityEngine::setNAdjClqCompressionBound(const int bound ){
  if ( qgraph != 0 ) { 
    return qgraph->setNAdjClqCompressionBound( bound );
  } else { 
    return false;
  }
}
*/

SPINDLE_END_NAMESPACE

#endif // ifndef MIN_PRIORITY_ORDERING_H_
