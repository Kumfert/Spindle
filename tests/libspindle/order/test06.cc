//
// src/order/drivers/tests/test06.cc -- test MinPriorityEngine
//
// $Id: test06.cc,v 1.2 2000/02/18 01:32:09 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//


#include <iostream.h>

#define DEBUG_TRACE true
#define DEBUG_OUT cout

#include "spindle/ArrayBucketSorter.h"
#include "spindle/MinPriorityEngine.h"
#include "samples/small_mesh.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#define __FUNC__ "int main"
int main() {
  FENTER;

  //  SpindleSystem::setTraceLevel( SPINDLE_TRACE_FLOW | SPINDLE_TRACE_DEBUG );
  SpindleSystem::setTraceFile( stdout );

  FCALL;
  Graph graph(nvtxs, (const int*) Aptr, (const int*)Aind);
  graph.validate();
  if ( !graph.isValid() ) {
    cerr << "Error: could not finalize graph " << endl;
    exit(-1);
  }

  FCALL;
  MinPriorityEngine ordering(&graph);
  if (DEBUG_TRACE) { cout << "Created Ordering" << endl; }
  
  //   ordering.setDegreeScheme( ordering.EXTERNAL_DEGREE );

  FCALL ordering.execute();
  if (DEBUG_TRACE) { cout << "Executing Ordering" << endl; }

  PermutationMap perm(9);
  perm.getOld2New().import( ordering.lendQuotientGraph().getOld2New(), 9);
  perm.validate();
  if ( perm.isValid() ) {
    const int * new2Old = perm.getNew2Old().lend();
    
    for (int i=0; i<9; i++) {
      cout << "perm[" << i << "] = " << new2Old[i] << endl;
    }
  } else {
    cout << "ERROR: PermutationMap is not valid" << endl;
  }
  FRETURN(0);
}
