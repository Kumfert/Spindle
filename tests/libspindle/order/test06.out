Created Ordering
Executing Ordering
perm[0] = 6
perm[1] = 0
perm[2] = 2
perm[3] = 8
perm[4] = 5
perm[5] = 3
perm[6] = 4
perm[7] = 7
perm[8] = 1
